tests/components/imap/test_coordinator.py

```diff
@@ -0,0 +1,197 @@
+    # Make sure we have had one update (when polling)
+    async_fire_time_changed(hass, utcnow() + timedelta(seconds=5))
+    await hass.async_block_till_done()
+    state = hass.states.get("sensor.imap_email_email_com")
+    # we should have received one message
+    assert state is not None
+    assert state.state == "1"
+
+    # cleanup event listener
+    handler()
+
+    # we should have received one event
+    event_called.assert_called_once()
+    data: dict[str, Any] = event_called.call_args[0][0].data
+    assert data["server"] == "imap.server.com"
+    assert data["username"] == "email@email.com"
+    assert data["search"] == "UnSeen UnDeleted"
+    assert data["folder"] == "INBOX"
+    assert data["sender"] == "john.doe@example.com"
+    assert data["subject"] == "Test subject"
+    assert data["text"]
+
+
+@pytest.mark.parametrize("imap_capabilities", [{"IDLE"}, set()], ids=["push", "poll"])
+@pytest.mark.parametrize(
+    ("imap_login_state", "success"), [(AUTH, True), (NONAUTH, False)]
+)
+async def test_initial_authentication_error(
+    hass: HomeAssistant, mock_imap_protocol: MagicMock, success: bool
+) -> None:
+    """Test authentication error."""
+    config_entry = MockConfigEntry(domain=DOMAIN, data=MOCK_CONFIG)
+    config_entry.add_to_hass(hass)
+    assert await hass.config_entries.async_setup(config_entry.entry_id) == success
+    await hass.async_block_till_done()
+
+
+@pytest.mark.parametrize("imap_capabilities", [{"IDLE"}, set()], ids=["push", "poll"])
+@pytest.mark.parametrize(
+    ("imap_select_state", "success"), [(AUTH, False), (SELECTED, True)]
+)
+async def test_initial_invalid_folder_error(
+    hass: HomeAssistant, mock_imap_protocol: MagicMock, success: bool
+) -> None:
+    """Test receiving a message successfully."""
+    config_entry = MockConfigEntry(domain=DOMAIN, data=MOCK_CONFIG)
+    config_entry.add_to_hass(hass)
+    assert await hass.config_entries.async_setup(config_entry.entry_id) == success
+    await hass.async_block_till_done()
+
+
+@pytest.mark.parametrize("imap_capabilities", [{"IDLE"}, set()], ids=["push", "poll"])
+@pytest.mark.parametrize("imap_search", [BAD_SEARCH_RESPONSE])
+@pytest.mark.parametrize(
+    ("exception", "error_message"),
+    [
+        (InvalidAuth, "Username or password incorrect, starting reauthentication"),
+        (InvalidFolder, "Selected mailbox folder is invalid"),
+    ],
+)
+async def test_late_authentication_or_invalid_folder_error(
+    hass: HomeAssistant,
+    caplog: pytest.LogCaptureFixture,
+    mock_imap_protocol: MagicMock,
+    imap_capabilities: set[str],
+    exception: InvalidAuth | InvalidFolder,
+    error_message: str,
+) -> None:
+    """Test authentication error after search was failed."""
+    config_entry = MockConfigEntry(domain=DOMAIN, data=MOCK_CONFIG)
+    config_entry.add_to_hass(hass)
+
+    if imap_capabilities == set():
+        # Avoid first refresh when polling to avoid a failing entry setup
```

Comment:
If we don't want to fail the setup, apply the side effect later, after the setup.

We don't want to have conditions or patch the coordinator methods in the tests.